From 4a8bfe906e415ae80958ff664ad642265d2be3c3 Mon Sep 17 00:00:00 2001
From: Prarit Bhargava <prarit@redhat.com>
Date: Thu, 28 Aug 2014 14:22:51 -0400
Subject: [cpufreq] distinguish drivers that do asynchronous notifications

Message-id: <1409235879-1283-78-git-send-email-prarit@redhat.com>
Patchwork-id: 89394
O-Subject: [RHEL7.1 PATCH BZ 1134639 077/185] cpufreq: distinguish drivers that do asynchronous notifications
Bugzilla: 1134369
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Mark Langsdorf <mlangsdo@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1134639

commit 7dbf694db6ac7c759599316d50d7050efcbd512a
Author: Viresh Kumar <viresh.kumar@linaro.org>
Date:   Tue Oct 29 18:56:06 2013 +0530

    cpufreq: distinguish drivers that do asynchronous notifications

    There are few special cases like exynos5440 which doesn't send POSTCHANGE
    notification from their ->target() routine and call some kind of bottom halves
    for doing this work, work/tasklet/etc.. From which they finally send POSTCHANGE
    notification.

    Its better if we distinguish them from other cpufreq drivers in some way so that
    core can handle them specially. So this patch introduces another flag:
    CPUFREQ_ASYNC_NOTIFICATION, which will be set by such drivers.

    This also changes exynos5440-cpufreq.c and powernow-k8 in order to set this
    flag.

    Acked-by: Amit Daniel Kachhap <amit.daniel@samsung.com>
    Acked-by: Kukjin Kim <kgene.kim@samsung.com>
    Signed-off-by: Viresh Kumar <viresh.kumar@linaro.org>
    Signed-off-by: Rafael J. Wysocki <rafael.j.wysocki@intel.com>

Cc: Lenny Szubowicz <lszubowi@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/cpufreq/powernow-k8.c b/drivers/cpufreq/powernow-k8.c
index 8d9f96c..c771db9 100644
--- a/drivers/cpufreq/powernow-k8.c
+++ b/drivers/cpufreq/powernow-k8.c
@@ -1204,6 +1204,7 @@ out:
 }
 
 static struct cpufreq_driver cpufreq_amd64_driver = {
+ .flags  = CPUFREQ_ASYNC_NOTIFICATION,
  .verify  = cpufreq_generic_frequency_table_verify,
  .target_index = powernowk8_target,
  .bios_limit = acpi_processor_get_bios_limit,
diff --git a/include/linux/cpufreq.h b/include/linux/cpufreq.h
index 8c76cce..72fd7e7 100644
--- a/include/linux/cpufreq.h
+++ b/include/linux/cpufreq.h
@@ -284,6 +284,13 @@ struct cpufreq_driver {
  */
 #define CPUFREQ_HAVE_GOVERNOR_PER_POLICY (1 << 3)
 
+/*
+ * Driver will do POSTCHANGE notifications from outside of their ->target()
+ * routine and so must set cpufreq_driver->flags with this flag, so that core
+ * can handle them specially.
+ */
+#define CPUFREQ_ASYNC_NOTIFICATION  (1 << 4)
+
 int cpufreq_register_driver(struct cpufreq_driver *driver_data);
 int cpufreq_unregister_driver(struct cpufreq_driver *driver_data);
 
-- 
1.7.1